rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,131 @@
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
+  configure image signing with `cosign` as the signing mechanism, specifying
+  the secret reference to get the private key and passphrase. The user can also
+  specify signature annotations as part of configuration:
+```yaml
+cosign:
+    secretRef: <secret-name> # private key and passphrase to sign images
+    annotations: # optional map of key-value pairs the user wants to add to images
+      key1: value1
+      key2: value2
+```
+`cosign` will apply this optional map of annotations to the signature payload,
+which can then be verified with `cosign verify`. Verification is out of scope
+for this RFC.
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it to the registry where the image is
+  located, using the same credentials that were used to push the image. This
+  flow must happen after the image has been pushed to the registry.
+
+- If `cosign` fails to sign an image, the build should fail and output an error
+  message specifying the failure.
+
+- Whenever `kpack` signs an image using `cosign`, it should add these
+  annotations:
+  - Build number.
+  - Cluster builder used.
+  - Buildpacks involved in the build process.
+  - Stack container image names and digests.
+  - Build time.
+
+### Key generation and storage
+
+`cosign` can use file-based keys, key management systems or hardware keys to
+ingest private keys, as well as a keyless flow that requires integration with
+[Rekor](https://github.com/sigstore/rekor).
+
+`kpack` should not generate the keys used for signing and verification. The user
+should pass it in using one of the mechanisms supported by `cosign`. In this RFC,
+we suggest that the first mechanism implemented be the use of Kubernetes
+`Secret`s for sending private key and passphrase into the build pod:
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cosign-key
+type: Opaque
+data:
+  cosign-privatekey: # base64 encoded private key here
```

Comment:
These key secret key names are hardcoded? Is there any prior art to this?

Reply:
We called the private key name `cosign-privatekey` because the SSH key secret type uses the name `ssh-privatekey` as its key field name.  Cosign has its own key format so we thought it was important to call it out in the key field name.  e.g.: if someone used `ssh-keygen` to make an X509 or ECDSA private key and passed it to `cosign` then `cosign` would fail.  At this time, it only accepts keys made with `cosign generate-key-pair`.

Reply:
Thanks for the explanation!